src/azure-cli/azure/cli/command_modules/role/tests/util.py

```diff
@@ -29,3 +30,17 @@ def retry(func, sleep_duration=ROLE_COMMAND_SLEEP_DURATION, max_retry=ROLE_COMMA
 def escape_apply_kwargs(val):
     """Replace {} as {{}} so that val is preserved after _apply_kwargs."""
     return val.replace('{', "{{").replace('}', "}}")
+
+
+class MSGraphUpnReplacer(RecordingProcessor):
+    """Replace UPN with encoded #
+    """
+    def __init__(self, test_name, mock_name):
+        self.test_name = test_name
+        self.test_name_encoded = test_name.replace('#', '%23')
+        self.mock_name = mock_name
+        self.mock_name_encoded = mock_name.replace('#', '%23')
+
+    def process_request(self, request):
+        request.uri = request.uri.replace(self.test_name_encoded, self.mock_name_encoded)
+        return request
```

Comment:
The old `AADGraphUserReplacer` only replaces the domain part

https://github.com/Azure/azure-cli/blob/427fc4d331526541dea7143849a9fa642019f8fe/src/azure-cli/azure/cli/command_modules/role/tests/latest/test_graph.py#L381

https://github.com/Azure/azure-cli/blob/8fc09120c6dd40bf43722c0c53b27a3279b4214c/src/azure-cli-testsdk/azure/cli/testsdk/utilities.py#L160-L168

and leaves the object name renaming job to `GeneralNameReplacer`. This works fine for object names without special characters, but this doesn't work in our current case since our object name contains special characters and is encoded, not only the domain part.